src/V3DfgDfgToAst.cpp

```diff
@@ -0,0 +1,318 @@
+    // Also builds the m_canonVars map as a side effect.
+    AstVar* getCanonicalVar(const DfgVar* vtxp) {
+        // Variable only read by DFG
+        if (!vtxp->driverp()) return vtxp->varp();
+
+        // Look up map
+        const auto it = m_canonVars.find(vtxp->varp());
+        if (it != m_canonVars.end()) return it->second;
+
+        // Not known yet, compute it (for all vars from the same driver)
+        std::vector<const DfgVar*> varps;
+        vtxp->driverp()->forEachSink([&](const DfgVertex& vtx) {
+            if (const DfgVar* const varVtxp = vtx.cast<DfgVar>()) varps.push_back(varVtxp);
+        });
+        UASSERT_OBJ(!varps.empty(), vtxp, "The input vtxp->varp() is always available");
+        std::stable_sort(varps.begin(), varps.end(), [](const DfgVar* ap, const DfgVar* bp) {
+            if (ap->hasExtRefs() != bp->hasExtRefs()) return ap->hasExtRefs();
+            const FileLine& aFl = *(ap->fileline());
+            const FileLine& bFl = *(bp->fileline());
+            if (const int cmp = aFl.operatorCompare(bFl)) return cmp < 0;
+            return ap->varp()->name() < bp->varp()->name();
+        });
+        AstVar* const canonVarp = varps.front()->varp();
+
+        // Add results to map
+        for (const DfgVar* const varp : varps) m_canonVars.emplace(varp->varp(), canonVarp);
+
+        // Return it
+        return canonVarp;
+    }
+
+    // Given a DfgVertex, return an AstVar that will hold the value of the given DfgVertex once we
+    // are done with converting this Dfg into Ast form.
+    AstVar* getResultVar(const DfgVertex* vtxp) {
+        const auto pair = m_resultVars.emplace(vtxp, nullptr);
+        AstVar*& varp = pair.first->second;
+        if (pair.second) {
+            // If this vertex is a DfgVar, then we know the variable. If this node is not a DfgVar,
+            // then first we try to find a DfgVar driven by this node, and use that, otherwise we
+            // create a temporary
+            if (const DfgVar* const thisDfgVarp = vtxp->cast<DfgVar>()) {
+                // This is a DfgVar
+                varp = getCanonicalVar(thisDfgVarp);
+            } else if (const DfgVar* const sinkDfgVarp = vtxp->findSink<DfgVar>()) {
+                // We found a DfgVar driven by this node
+                varp = getCanonicalVar(sinkDfgVarp);
+            } else {
+                // No DfgVar driven by this node. Create a temporary.
+                const string name = m_tmpNames.get(vtxp->hash(m_hashCache).toString());
+                // TODO: Do we ever need to make this VSigning::SIGNED?
+                AstNodeDType* const dtypep = v3Global.rootp()->findBitDType(
+                    vtxp->width(), vtxp->width(), VSigning::UNSIGNED);
+                varp = new AstVar{vtxp->fileline(), VVarType::MODULETEMP, name, dtypep};
+                // Add temporary AstVar to containing module
+                m_modp->addStmtp(varp);
```

Comment:
I wonder if we ever need to make a signed temporary here?
This is used to render code that has common sub-expressions:
```systemverilog
  assign x = a + b - p;
  assign y = a + b - q;
```
At the end of DFG this will be rendered as 
```systemverilog
   wire logic [width-1:0] tmp = a + b;
   assign x = tmp + p;
   assign y = tmp + q;
```
So the question is, is it safe for tmp to always be unsigned?

I could restrict AstToDfg so anything with a signed term is not optimized, but there is also `$signed` and friends that might introduce more. It would be ideal to handle these.

Reply:
I think the temp should be able to keep the signing of the expression to avoid trouble, while I can't think of things that might break this is what all other temporary insertions do.  I don't see what would prevent optimizing signing - the rules should be setup to allow those cases - generally the Ast types are different for any cases where signing matters.

Reply:
> generally the Ast types are different for any cases where signing matters

That is my understanding. So can I just set the temporary signed, if the DFG vertex that generates it is a signed operation?

Reply:
Correct, or generally just take the dtype entirely from the expression.

Reply:
> or generally just take the dtype entirely from the expression.

Dfg vertices do not have dtypes, just widths, hence my peril :) (and would prefer not to add one as it's not really relevant, and hard/verbose to keep updaed, the vertex type should fully determine what the result is, and like AstNodeMath, the Dfg vertex type determines the interpretation of the input, otherwise everything is just a packed vector of bits)

Reply:
I think I have reasoned myself into believing that always unsigned temporaries are OK, because we only add a temporary for intermediate expressions that are read by other expressions (i.e.: there will be no readers other than those expressions), and expressions (past V3Width) interpret their inputs based on the expression node type, not the operand types.

Does this sound reasonable?

(this also made me realise that we should not merge signed and unsigned *variables* in this pass, so I will fix that).

Reply:
> (this also made me realise that we should not merge signed and unsigned variables in this pass, so I will fix that).

Looking more into this, even this should not matter post V3Width and some early front-end passes (basically where we need to determine whether to sign or zero extend soemthing, this optimization is after those). Signed and unsigned variables are emitted the same in the output and both have unused upper bits cleared by V3Clean on assignment, so again I think always unsigned temporaries here are OK.